src/node/configuration_tracker.h

```diff
@@ -0,0 +1,413 @@
+#include "enclave/rpc_sessions.h"
+#include "kv/kv_types.h"
+#include "kv/store.h"
+#include "node/config.h"
+#include "node/entities.h"
+#include "node/nodes.h"
+#include "node/rpc/node_call_types.h"
+#include "node/rpc/serdes.h"
+#include "rpc/serialization.h"
+
+using namespace ccf;
+
+namespace aft
+{
+  using Index = uint64_t;
+
+  struct NodeState
+  {
+    kv::Configuration::NodeInfo node_info;
+
+    // the highest index sent to the node
+    Index sent_idx;
+
+    // the highest matching index with the node that was confirmed
+    Index match_idx;
+
+    // Flag indicating whether this node is still catching up; before that it's
+    // not an eligible voter.
+    bool catching_up;
+
+    NodeState() = default;
+
+    NodeState(
+      const kv::Configuration::NodeInfo& node_info_,
+      Index sent_idx_,
+      Index match_idx_ = 0,
+      bool catching_up_ = true) :
+      node_info(node_info_),
+      sent_idx(sent_idx_),
+      match_idx(match_idx_),
+      catching_up(catching_up_)
+    {}
+  };
+
+  class ConfigurationTracker
+  {
+  public:
+    const NodeId& node_id;
+    ConsensusType consensus_type;
+    std::shared_ptr<kv::AbstractStore> store;
+    std::shared_ptr<enclave::RPCSessions> rpcsessions;
+    std::shared_ptr<enclave::RPCMap> rpc_map;
+    crypto::KeyPairPtr node_sign_kp;
+    const crypto::Pem& node_cert;
+    std::unordered_map<ccf::NodeId, aft::NodeState>& nodes;
+    std::list<kv::Configuration> configurations;
+
+  public:
+    ConfigurationTracker(
+      const NodeId& node_id_,
+      ConsensusType consensus_type_,
+      std::unordered_map<ccf::NodeId, aft::NodeState>& nodes_,
+      std::shared_ptr<kv::AbstractStore> store_,
+      std::shared_ptr<enclave::RPCSessions> rpcsessions_,
+      std::shared_ptr<enclave::RPCMap> rpc_map_,
+      const crypto::KeyPairPtr& node_sign_kp_,
+      const crypto::Pem& node_cert_) :
+      node_id(node_id_),
+      consensus_type(consensus_type_),
+      store(store_),
+      rpcsessions(rpcsessions_),
+      rpc_map(rpc_map_),
+      node_sign_kp(node_sign_kp_),
+      node_cert(node_cert_),
+      nodes(nodes_)
+    {
+      if (!node_sign_kp_)
+        LOG_INFO_FMT("NO SIGNING KEY!");
+      if (!rpc_map)
+        LOG_INFO_FMT("NO RPC MAP!");
+    }
+
+    ~ConfigurationTracker() {}
+
+    // Examine all configurations that are followed by a globally committed
+    // configuration.
+    bool commit(Index idx)
+    {
+      LOG_TRACE_FMT("Configurations: commit {}!", idx);
+
+      bool r = false;
+      while (true)
+      {
+        auto conf = configurations.begin();
+        if (conf == configurations.end())
+          break;
+
+        auto next = std::next(conf);
+        if (next == configurations.end())
+          break;
+
+        if (idx < next->idx)
+          break;
+
+        configurations.pop_front();
+        r = true;
+      }
+
+      LOG_TRACE_FMT("Configurations: {}", to_string());
+
+      return r;
+    }
+
+    bool rollback(Index idx)
```

Comment:
This is nit: but i find this API confusing, returning false to me implies that the rollback failed not that there was nothing to rollback